macros/contexts/contextNondecimalBase.pl

```diff
@@ -0,0 +1,273 @@
+
+=head1 NAME
+
+contextHex.pl - Implements a MathObject class and context for integers
+                in hexadecimal notation.
+
+=head1 DESCRIPTION
+
+This context implements a Hex object that works like a Real, but
+where you enter numbers in hexadecimal, and they display in hexadecimal.
+You can perform the usual numeric operations (addition, subtraction,
+etc.), but division is integer division (so 7/3 = 2).  The context defines
+the bitwise operators &, |, ^, >>, <<, and ~ (for bitwise and, or,
+exclusive or, shift-right, shift-left, and one's complement not).  You can
+apply these operations within your PG code to variables that store Hex
+objects.  Remember that you can also obtain Perl reals via hex notation,
+for example, 0x1A.
+
+To use hexadecimal MathObjects, first load the contextHex.pl file:
+
+	loadMacros("contextHex.pl");
+
+and then select the appropriate context -- one of the following:
+
+	Context("Hex");
+	Context("LimitedHex");
+
+The latter only allows the student to enter hexadecimal literals (not
+expressions), or lists of hexadecimal numbers.  The former allows
+expression involving numeric operations and bitwise operations.
+
+Once one of these contexts is selected, all the nummbers parsed by
+MathObjects will be considered to be in hexadecimal, so
+
+        $n = Compute('10');
+
+produces the hexadecimal number 10 (decimal 16).  You could also obtain
+a hex MathObject using
+
+        $n = Hex(0x10);
+
+Once you have such a value, use
+
+        ANS($n->cmp)
+
+to get an answer checker for the number.  You can also perform numeric or
+bitwise operations on the value, as in
+
+        $m = $n + 0xE3;
+        $N = $n << 2;     # shifts n to the left 2 (binary) places,
+                          #   $N = Hex(0x40) when $n = Hex(0x10)
+
+=cut
+
+sub _contextNondecimalBase_init {
+	context::NondecimalBase::Init();
+	sub setBase     { context::NondecimalBase::setBase(@_); }
+	sub convertBase { context::NondecimalBase::convert(@_); }
+}
+
+###########################################################################
+
+package context::NondecimalBase;
+
+# defines the base for the context.
+our $base = 10;
+#
+#  The standard digits, pre-built so it doesn't have to be done each time the conversion is called
+#
+our $digits16 = [ '0' .. '9', 'A' .. 'F' ];
+our $digit16  = { map { ($digits16->[$_], $_) } (0 .. scalar(@$digits16) - 1) };
+
+use Data::Dumper;
+#
+#  Initialize the contexts and make the creator function.
+#
+sub Init {
+	my $context = $main::context{NondecimalBase} = Parser::Context->getCopy("Numeric");
+	$context->{name} = 'NondecimalBase';
+	$context->{parser}{Number} = 'context::NondecimalBase::Number';
+	# $context->{value}{Real} = 'context::Hex::Hex';
+	$context->{pattern}{number} = '[0-9A-F]+';
+	$context->functions->disable('All');
+	$context->operators->remove('^');
```

Comment:
This was removed because it was redefined below (as exclusive or).  If you are not going to add the bitwise operations, you can leave `^` as is, and not remove it.